Fix -fno-unsafe-math-optimizations behavior

andykaylor opened the pull request:
This changes the handling of -fno-unsafe-fp-math to stop having that
option imply -ftrapping-math. In gcc, -fno-unsafe-math-optimizations sets
-ftrapping-math, but that dependency is based on the fact the
-ftrapping-math is enabled by default in gcc. Because clang does not
enable -ftrapping-math by default, there is no reason for
-fno-unsafe-math-optimizations to set it.

On the other hand, -funsafe-math-optimizations continues to imply
-fno-trapping-math because this option necessarily disables strict
exception semantics.

This fixes https://github.com/llvm/llvm-project/issues/87523

arsenm commented on clang/test/Driver/fast-math.c:
```diff
@@ -271,30 +271,32 @@
 
 // CHECK-NO-UNSAFE-MATH: "-cc1"
 // CHECK-NO-UNSAFE-MATH-NOT: "-funsafe-math-optimizations"
-// CHECK-NO_UNSAFE-MATH-NOT: "-mreassociate"
```
So this test was full of broken test checks?

zahiraam replied:
It looks like the lines with `_` were just ignored! That's bad.

arsenm commented on clang/test/Driver/fast-math.c:
```diff
@@ -318,12 +320,12 @@
-// CHECK-NO-REASSOC-NO_UNSAFE-MATH-NOT: "-funsafe-math-optimizations"
-// CHECK-NO-REASSOC-NO_UNSAFE-MATH-NOT: "-mreassociate"
-// CHECK-NO-REASSOC-NO_UNSAFE-MATH-NOT: "-funsafe-math-optimizations"
+// CHECK-NO-REASSOC-NO-UNSAFE-MATH-NOT: "-funsafe-math-optimizations"
+// CHECK-NO-REASSOC-NO-UNSAFE-MATH-NOT: "-mreassociate"
+// CHECK-NO-REASSOC-NO-UNSAFE-MATH-NOT: "-funsafe-math-optimizations"
```
I don't think this testing technique works. All of these flags would be printed on the same line as cc1. You would need to use CHECK-SAME-NOT, but I'm not sure you can actually compose -SAME and -NOT checks 

-NOT checks are hazardous this way, can we positively check for the flags that are emitted.

andykaylor replied:
I'd be happy to make that change. I was a little lazy with this test. I just wanted something that tested the change I was making, and then the rest of what I did was just what seemed necessary to make the test pass while still testing the things it meant to test. The whole thing seemed a bit of a mess. I can clean it up some with a bit more effort.

zahiraam replied:
Yes, I think that should work!

andykaylor commented:
I've made some minor changes to clean up the LIT test checks for this PR. I started to clean up the entire test, but it quickly became obvious that I should put that into a separate PR.

I will submit a new PR to clean up the entire fast-math.c test. I have an idea that I think will work, hard-coding the order of the math-related flags and using a separate prefix for each one plus negative checks. Then the test can just use "--check-prefixes=..." to put together the combinations that are expected. I think that will be more maintainable and more readable than trying to group checks in the more-or-less arbitrary way the test currently does.

andykaylor commented:
I've made extensive changes to the fast-math.c test here: https://github.com/llvm/llvm-project/pull/89687

That update isn't intended to change any existing behavior. A few things worked differently than I expected, but I left the checks consistent with the current behavior and added comments where I thought the current behavior was incorrect.

Commit:
Fix -fno-unsafe-math-optimizations behavior

This changes the handling of -fno-unsafe-fp-math to stop having that
option imply -ftrapping-math. In gcc, -fno-unsafe-math-optimizations sets
-ftrapping-math, but that dependency is based on the fact the
-ftrapping-math is enabled by default in gcc. Because clang does not
enabled -ftrapping-math by default, there is no reason for
-fno-unsafe-math-optimizations to set it.

On the other hand, -funsafe-math-optimizations continues to imply
-fno-trapping-math because this option necessarily disables strict
exception semantics.

This fixes https://github.com/llvm/llvm-project/issues/87523

andykaylor commented:
ping for review

zahiraam commented:
LGTM! Thanks.
